aspnetcore/fundamentals/middleware/rate-limit/samples/WebRateLimitAuth/Program.cs

```diff
@@ -0,0 +1,196 @@
+#define FIRST // FIRST ADMIN
+#if NEVER
+#elif FIRST
+// <snippet_1>
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.RateLimiting;
+using Microsoft.EntityFrameworkCore;
+using System.Net;
+using System.Threading.RateLimiting;
+using WebRateLimitAuth.Data;
+
+var builder = WebApplication.CreateBuilder(args);
+
+var connectionString = builder.Configuration.GetConnectionString("DefaultConnection") ??
+    throw new InvalidOperationException("Connection string 'DefaultConnection' not found.");
+builder.Services.AddDbContext<ApplicationDbContext>(options =>
+    options.UseSqlServer(connectionString));
+builder.Services.AddDatabaseDeveloperPageExceptionFilter();
+
+builder.Services.AddDefaultIdentity<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
+    .AddEntityFrameworkStores<ApplicationDbContext>();
+builder.Services.AddRazorPages();
+
+var app = builder.Build();
+
+if (app.Environment.IsDevelopment())
+{
+    app.UseMigrationsEndPoint();
+}
+else
+{
+    app.UseExceptionHandler("/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+// <snippet>
+app.UseAuthentication();
+app.UseAuthorization();
+
+var userPolicyName = "user";
+
+var options = new RateLimiterOptions()
+    .AddPolicy<string>(userPolicyName, context =>
+    {
+        if (!context.User?.Identity?.IsAuthenticated ?? false)
+        {
+            var username = "anonymous user";
+
+            return RateLimitPartition.CreateSlidingWindowLimiter<string>(username,
+                  key => new SlidingWindowRateLimiterOptions(
+                  permitLimit: 2,
+                  queueProcessingOrder: QueueProcessingOrder.OldestFirst,
+                  queueLimit: 0,
+                  window: TimeSpan.FromSeconds(30),
+                  segmentsPerWindow: 1
+                ));
+        }
+        else
+        {
+            return RateLimitPartition.CreateNoLimiter<string>(string.Empty);
+        }
+    });
+
+options.GlobalLimiter = PartitionedRateLimiter.Create<HttpContext, IPAddress>(context =>
```

Comment:
@pentp  wrote
> It would be more appropriate to use a TokenBuckedRateLimiter here and it would also provide an example of a different kind of rate limiter.

I changed this to `TokenBuckedRateLimiter`. Because it uses the `remoteIPaddress` for the partition key, we'll have too many partitions. Shouldn't this use a fixed partition key for non-loopback IPs? Why do we need all those partition keys when the rate limiter is the same?

Reply:
> Shouldn't this use a fixed partition key for non-loopback IPs? Why do we need all those partition keys when the rate limiter is the same?

It depends on how we want the limiter to work. Yes there would be too many partitions but in this scenario each IP would be limited separately. 

If there's only one partition key for non-loopback IPs. One IP could send too many requests and block the system for all other IPs.